### PR TITLE
Add OLMo model family

### LINKED ISSUE #29885: Add OLMo family
### Model description

OLMo is a series of Open Language Models designed to enable the science of language models. The OLMo models are trained on the [Dolma](https://huggingface.co/datasets/allenai/dolma) dataset. OLMo is releasing all code, checkpoints, logs (coming soon), and details involved in training these models.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

Authored by Allen Institute for AI (HF org: [allenai](https://huggingface.co/allenai))

Implementation: https://github.com/allenai/OLMo/tree/main/olmo
HF Hub: [OLMo-1B](https://huggingface.co/allenai/OLMo-1B), [OLMo-7B](https://huggingface.co/allenai/OLMo-7B), [OLMo-7B-Twin-2T](https://huggingface.co/allenai/OLMo-7B-Twin-2T)

Weights in HF formats (`.safetensors` and `.bin`) can be found in the respective HF Hub page. Each of these repos has branches containing intermediate checkpoints.
Weights in the original OLMo format can be retrieved following the instructions on the Github page.

### PR DESCRIPTION
**Pull request by 2015aroras**
# What does this PR do?

This PR adds the OLMo model family to transformers. A base `OLMoModel` and a casual LM `OLMoForCausalLM` are implemented. The models are already present in HF Hub (e.g. [allenai/OLMo-7B](https://huggingface.co/allenai/OLMo-7B)), and this implementation is compatible with the checkpoints in the Hub.

UPDATE: The current version of the PR is not compatible with the old OLMo models in HF Hub. New models have been uploaded to HF Hub (e.g. [allenai/OLMo-7B-hf](https://huggingface.co/allenai/OLMo-7B-hf)) to support this PR.

Fixes #29885

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by 2015aroras** [comment]
@ArthurZucker and @younesbelkada for general guidance and/or review, since this is a PyTorch text model.

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for submitting the PR!
The code seems in a very research state, let's try to get it to `transformers` philosophy. 
My main query would be: what are the main differences with `MixtralMoe` ? (Llama + MoE) ? Only include differences for trained checkpoints (I see both alibi and rotary, all of them use both? Or only one was kept?) 

The fast tokenizer also looks very similar to GptNeoX / Bloom one, any reason to have a new one?

**Comment by 2015aroras** [comment]
> Hey! Thanks for submitting the PR! The code seems in a very research state, let's try to get it to `transformers` philosophy. My main query would be: what are the main differences with `MixtralMoe` ? (Llama + MoE) ? Only include differences for trained checkpoints (I see both alibi and rotary, all of them use both? Or only one was kept?)
> 
> The fast tokenizer also looks very similar to GptNeoX / Bloom one, any reason to have a new one?

In a general sense, I tried to stay as faithful to OLMo as I could while following general `transformers` principles (all model code in 1 file, use "Copied from" with re-usable logic, etc.).

Regarding tokenization, OLMo adds an eos token to the end of inputs that don't have them. The GptNeoX and Bloom ones don't. Llama's tokenizer does, but it also has a slow tokenizer and I couldn't figure out how to convert the fast OLMo tokenizer to a slow `sentencepiece` one without reinventing the wheel. My current implementation copies a lot from the Llama fast tokenizer.

In terms of the model, OLMo doesn't have MoE whereas `MixtralMoe` does. I would have to figure out if I could use `MixtralMoe` while avoiding the MoE functionality. OLMo is more directly map-able to Llama; the model architectures mostly map 1 to 1, and the main discrepancies in output come from the 2 models using different precision for some operations.

The official OLMo models use rope and at most 1 of rope and alibi is used in a run (i.e. no alibi), but we hope to release checkpoints from our experiments at some point. Some of those would use alibi. I think it would be nice if transformers could support those checkpoints too.

@dirkgr FYI

**Comment by ArthurZucker** [comment]
Hey! Sorry I might have misunderstood (thought I saw some MoE things here). 
Alright when I mean research I mean there are a lot of things that we never do in transformers. 
1. Tokenizer, we can update bloom / Gpneox to support this option as IMO it makes a lot of sense! No need for a slow one 😉 
2. The modeling code does not seem to include a lot of `# Copied from`, would recommend you to look at `Gemma` modeling for a good example! Basically we want to avoid all the autocasting code when it is not necessary. we do checkpointing a bit differently and we want to have the least number of code paths as possible. 
3. For the config, all of the different enums for layer norms and act fn, we already have the `from ...activations import ACT2FN
` and the `ALL_LAYER_NORMS` which should be used. 
4. `BufferCache` is not really something we use, and I am not sure we had issue with the community after the last Llama fixes!
5. `k_norm` and `q_norm` are used in `Persimmon` so a good place to copy from as well!
6. `_scaled_dot_product_attention` should follow the classing we use! So xxxxSdpaAttention
7. if `clip_qkv` is always true let's not have a if else!
8. attention mask logic should use the `_update_causal_mask` from Llama and have a similar logic
9. We don't use `reset_parameters` or `device` let's try to get closer to Llama!
10. We don't incude `get_fsdp_wrap_policy` and fsdp specifics in transformers
11. Let's add some generation tests!

**Comment by 2015aroras** [comment]
Just want to clarify my thoughts before I try making changes, so I'm expressing my understanding below. Please correct me where I am wrong.

Regarding the tokenizer, it sounds like (if possible) I should update one of the existing tokenizers and then use that tokenizer directly (similar to how `mamba` appears to use `GPTNeoXTokenizerFast`).

The comments regarding modeling code suggest that I should aim to re-use existing code using `#Copied from` and any existing functions. More generally, I should aim to be more "transformers-like" than faithful to the original code. I should try to make the code run like OLMo (produce the same output) but look like transformers.

----

Regarding copying existing code vs being faithful, what should I do when some operation (say, RoPE) exists in transformers but is performed in a different precision (e.g. float32 vs bfloat16) in OLMo compared to the transformer code? Do I use the transformers code and disregard the "small" error, or write separate code that matches the original OLMo behavior?

**Comment by ArthurZucker** [comment]
> Assuming it were possible, the ideal solution would be for me to use some existing model and convert OLMo checkpoints to work with that model.

Yes! This way it's easier for all the community to use the model, and keeps the diff very clear and clean! 

For the tokenizer, if `GPTNeoXTokenizerFast` does not have `add_eos` or `add_bos`, we can add it but in a backward compatible way! 

> The comments regarding modeling code suggest that I should aim to re-use existing code using #Copied from and any existing functions. More generally, I should aim to be more "transformers-like" than faithful to the original code. I should try to make the code run like OLMo (produce the same output) but look like transformers.

Exactly 💯

**Comment by 2015aroras** [comment]
Due to the above discussion, I have redone the implementation. The key points are:

- For tokenizer, OLMo now uses `GPTNeoXTransformerFast`. I have modified the tokenizer to make adding `bos_token` or `eos_token` an option. Unfortunately, that tokenizer doesn't come with any tests...
- For model, I have reverted mostly to the code generated from [add-new-model-like with Llama](https://github.com/huggingface/transformers/pull/29890/commits/0afbd4a35925394ee4c9010e6bf985d87adac764). The main changes are changing the norm into the non-parametric version used by OLMo and removing auto-generated model classes like `OLMoForQuestionAnswering`. Should I add those classes back (I have not spent time trying to understand those classes yet)?

The existing checkpoints on HF Hub are not compatible with this implementation, so integration tests are currently broken. I can work on fixing that once we are happy with the approach/implementation. On a 1B checkpoint that I converted using the updated conversion script, the maximal logit difference is `~2.5e-5`.

**Comment by ArthurZucker** [comment]
Thanks! Pinging @molbap for the review!

**Comment by 2015aroras** [comment]
@molbap Gentle ping

**Review by molbap — COMMENTED** [review]
Hey, got a review out! I thnk the code is in a good state- I left some pointers around the places where it seems to break. If there is no change in the way `causal_mask` is used w.r.t. llama/gemma, I would advise to use the original `_update_causal_mask` method! 

Let me know when integration + tokenizers test are green on your end and I'll do another pass!

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py]: https://github.com/huggingface/transformers/pull/29890/commits/98db6a929eb7170846e34ac93de183cc457218ae Sorry, I missed a 'Copied from' here. `_update_causal_mask ` this is supposed to be identical to Llama. The only part of the OLMo model code that is supposed to be different is the non-parametric layer norm.

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py]: Refer to https://github.com/huggingface/transformers/pull/29890#discussion_r1554288624

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py]: Refer to https://github.com/huggingface/transformers/pull/29890#discussion_r1554288624

File: tests/models/olmo/test_tokenization_olmo.py
Comment (2015aroras) [reply tests/models/olmo/test_tokenization_olmo.py:1]: I believe that one test from `Mixin` with `['allenai/OLMo-1B']` instead of `'allenai/OLMo-1B'` was a bug with the test instead of my code. After merging an updated main, the error is gone.

I currently have `1 failed, 91 passed, 4 skipped, 7 warnings` when I run. The one failure (`OLMoTokenizationTest::test_tokenizer_integration`) is because the current OLMo-1B tokenizer config doesn't have `"add_eos_token": true`, since it is not yet updated. Are you seeing more failures?

File: tests/models/olmo/test_modeling_olmo.py
Comment (2015aroras) [reply tests/models/olmo/test_modeling_olmo.py:420]: Yep, I haven't sorted out the integration tests yet so that is just the auto-generated `Llama-hf` -> `OLMo-hf`.

**Comment by 2015aroras** [comment]
Do you have any recommendations for how to transition our checkpoints in HF Hub from the old style that relied on remote code to the new checkpoints from this PR? I can accordingly work on updating HF Hub and fixing the integration tests.

Other than integration tests that rely on a HF Hub checkpoint, everything else should be passing now.

**Comment by molbap** [comment]
@2015aroras For checkpoint adaptation, I'd take a look at [conversion scripts for gemma ](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma/convert_gemma_weights_to_hf.py). Overall not too far from what you are doing, load from config with correctly shaped parameters. Usually how I go about it is running the two implementations side-by-side in 2 different debuggers and look where the activations start to diverge, fixing the conversion script as I go.
For the tests, with the rebase it looks better now indeed! When the integration tests run we'll be able to ask for a core maintainer final approval :)

**Comment by 2015aroras** [comment]
Sorry, I should have been more specific about my concerns regarding transitioning our checkpoints in HF hub. My problem is that if I load the new-style OLMo checkpoints (i.e. OLMo checkpoints converted using `convert_olmo_weights_to_hf.py`), then it would break HF OLMo for people without the code from this PR. OLMo is already consumable using `OLMoForCausalLM.from_pretrained("OLMo-7B", trust_remote_code=True)`, but once I upload the new-style checkpoint their code will break until they update the HF library. How can I upload the new-style OLMo checkpoints without breaking people that are using older `transformers` versions?

It looks like Llama got around this issue by having separate `meta-llama/Llama-2-7b` and `meta-llama/Llama-2-7b-hf`, where the latter is "converted for the Hugging Face Transformers format". Is this approach acceptable?

**Comment by molbap** [comment]
Ah, I understand better now! Yes, we usually recommend to go for one repo on the hub = one model. In that case, it would be acceptable to add a -hf suffix to the new checkpoints on the hub!

**Comment by 2015aroras** [comment]
I have now put up "-hf" models as _private_ on the allenai HF Hub organization. Once I get the all clear internally, we will make them public. All the integration tests pass for me locally (they fail here because the repo is private).

We are hoping to release some new model checkpoints using the same architecture early next week, so it would be great if we could get this PR in quickly. Then we would be able to release models natively integrated with HF along with the rest of the release!

File: src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py
Comment (2015aroras) [inline src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:83]: I have switched OLMo to not add the eos token by default, since that is its behavior at inference time. Consequently the changes to this tokenizer are no longer strictly needed. There is no harm to keep them though, and the largest parts of the changes are copied from Llama anyways.

Comment (molbap) [reply src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:83]: yeah I don't think that's an issue!

File: src/transformers/models/llama/modeling_llama.py
Comment (molbap) [inline src/transformers/models/llama/modeling_llama.py:1114]: cc @ydshieh is this alright for the doctest?

File: src/transformers/models/olmo/configuration_olmo.py
Comment (molbap) [inline src/transformers/models/olmo/configuration_olmo.py:129]: is it wanted that default config do not set the `bos_token_id`?

**Comment by molbap** [comment]
I think the PR is in a good state. Thanks for adding this! pinging @ArthurZucker for a final review :)

File: src/transformers/models/olmo/configuration_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/configuration_olmo.py:129]: OLMo's config class (in the original Github) doesn't have a bos token and so no OLMo checkpoints use bos tokens to my knowledge. Thus setting no `bos_token_id` is intentional.

**Comment by 2015aroras** [comment]
@molbap @ArthurZucker All checks are passing! Now it's just a matter of reviews

File: src/transformers/models/auto/configuration_auto.py
Comment (ArthurZucker) [inline src/transformers/models/auto/configuration_auto.py:177]:
```suggestion
        ("olmo", "OlmoConfig"),
```
we need camel casing on all classes

File: src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:1]: LGTM

File: src/transformers/models/llama/modeling_llama.py
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1]: let's not modify an unrelated file as important as Llama! Remove the copied from and for the forward only! / use `# Ignore copy`

File: src/transformers/models/olmo/configuration_olmo.py
Comment (ArthurZucker) [inline src/transformers/models/olmo/configuration_olmo.py:79]: Let's get rid of that no?

File: src/transformers/models/olmo/modeling_olmo.py
Comment (ArthurZucker) [inline src/transformers/models/olmo/modeling_olmo.py:63]: this is copied from

File: src/transformers/models/olmo/modeling_olmo.py
Comment (ArthurZucker) [inline src/transformers/models/olmo/modeling_olmo.py:181]:
```suggestion
# Copied from transformers.models.llama.modeling_llama.rotate_half
def rotate_half(x):
    """Rotates half the hidden dims of the input."""
    x1 = x[..., : x.shape[-1] // 2]
    x2 = x[..., x.shape[-1] // 2 :]
    return torch.cat((-x2, x1), dim=-1)

```

**Review by ArthurZucker — COMMENTED** [review]
Good state!
- camel casing needs to be used on a ll classes!
- the main diff is `clamping`, it should not be an if else, because else it's a llama model 🤗

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py:360]: It's not used for the old OLMo model that is currently on HF Hub, but it is always used with the new OLMo model that we are releasing soon. We're going to have both on HF Hub in the long term, so I don't see a way around this.

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py:185]: [3c60cf3](https://github.com/huggingface/transformers/pull/29890/commits/3c60cf3a872de1d3f6dd338a370cb2bcc6378572) Used Llama instead, since that is the reference model for this implementation

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py:182]: [3c60cf3](https://github.com/huggingface/transformers/pull/29890/commits/3c60cf3a872de1d3f6dd338a370cb2bcc6378572)

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py:63]: [3c60cf3](https://github.com/huggingface/transformers/pull/29890/commits/3c60cf3a872de1d3f6dd338a370cb2bcc6378572)

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py]: [a037fca](https://github.com/huggingface/transformers/pull/29890/commits/a037fcac5e098cd22443531c618d42513b51b085)

File: tests/models/olmo/test_tokenization_olmo.py
Comment (2015aroras) [reply tests/models/olmo/test_tokenization_olmo.py:1]: [ec344c3](https://github.com/huggingface/transformers/pull/29890/commits/ec344c3af3332b9941df6dae3025f17da373fb42)

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py]: [4f062d7](https://github.com/huggingface/transformers/pull/29890/commits/4f062d76bc145aadede0cc405a0114aafe627ad7)

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py]: [4f062d7](https://github.com/huggingface/transformers/pull/29890/commits/4f062d76bc145aadede0cc405a0114aafe627ad7)

File: src/transformers/models/llama/modeling_llama.py
Comment (2015aroras) [reply src/transformers/models/llama/modeling_llama.py:1]: [826c91b](https://github.com/huggingface/transformers/pull/29890/commits/826c91ba7106eacfb94de8dfd1f7870559e08a95)

File: src/transformers/models/auto/configuration_auto.py
Comment (2015aroras) [reply src/transformers/models/auto/configuration_auto.py]: [c7f4e74](https://github.com/huggingface/transformers/pull/29890/commits/c7f4e7489836d43b1d6a34cf47eb1de1120d1961)

**Comment by 2015aroras** [comment]
Addressed all PR comments, and hopefully all checks will pass.

File: src/transformers/models/olmo/modeling_olmo.py
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py:325]: Just to be clear, the other main diff from Llama is the non-parametric layer norm.

Comment (ArthurZucker) [reply src/transformers/models/olmo/modeling_olmo.py:325]: Mmmm got it

**Review by ArthurZucker — APPROVED** [review]
Thanks for your great work! 🔥

**Pull request merged**
